pkg/images/images.go

```diff
@@ -0,0 +1,203 @@
+	DefaultAgentImageName        string = "agent"
+	DefaultClusterAgentImageName string = "cluster-agent"
+	OTelAgentBetaTag                    = "7.63.0-ot-beta-jmx"
+)
+
+// imageHasTag identifies whether an image string contains a tag suffix
+// Ref: https://github.com/distribution/distribution/blob/v2.7.1/reference/reference.go
+var imageHasTag = regexp.MustCompile(`.+:[\w][\w.-]{0,127}$`)
+
+// imageNameContainsTag return true if the image name contains a tag
+func imageNameContainsTag(name string) bool {
+	// The image name corresponds to a full image string
+	return imageHasTag.MatchString(name)
+}
+
+// Image represents a container image information
+type Image struct {
+	registry string
+	name     string
+	tag      string
+	isJMX    bool
+	isFIPS   bool
+}
+
+// NewImage return a new Image instance
+// Assumes that tag suffixes have already been trimmed and booleans updated accordingly
+func NewImage(registry, name, tag string, isJMX bool, isFIPS bool) *Image {
+	return &Image{
+		registry: registry,
+		name:     name,
+		tag:      tag,
+		isJMX:    isJMX,
+		isFIPS:   isFIPS,
+	}
+}
+
+func (i *Image) WithRegistry(registry string) {
+	if registry == "" {
+		return
+	}
+	i.registry = registry
+}
+
+func (i *Image) WithTag(tag string) {
+	if tag == "" {
+		return
+	}
+	i.tag = tag
+}
+
+func (i *Image) WithName(name string) {
+	if name == "" {
+		return
+	}
+	i.name = name
+}
+
+func (i *Image) WithJMX(isJMX bool) {
+	i.isJMX = isJMX
+}
+
+func (i *Image) WithFIPS(isFIPS bool) {
+	i.isFIPS = isFIPS
+}
+
+// GetLatestAgentImage return the latest stable agent release version
+func GetLatestAgentImage() string {
+	image := NewImage(DefaultImageRegistry, DefaultAgentImageName, AgentLatestVersion, false, false)
+	return image.ToString()
+}
+
+// GetLatestClusterAgentImage return the latest stable agent release version
+func GetLatestClusterAgentImage() string {
+	image := NewImage(DefaultImageRegistry, DefaultClusterAgentImageName, ClusterAgentLatestVersion, false, false)
+	return image.ToString()
+}
+
+// AssembleImage builds the image string based on ImageConfig and the registry configuration.
+func AssembleImage(imageSpec *v2alpha1.AgentImageConfig, registry string) string {
+	if imageNameContainsTag(imageSpec.Name) {
+		return imageSpec.Name
+	}
+
+	if registry == "" {
+		registry = DefaultImageRegistry
+	}
+
+	tag := imageSpec.Tag
+	// If JMXEnabled, then proactively trim JMX suffix to prevent duplicate suffixes
+	if imageSpec.JMXEnabled {
+		tag = strings.TrimSuffix(imageSpec.Tag, JMXTagSuffix)
+	}
+
+	img := NewImage(registry, imageSpec.Name, tag, imageSpec.JMXEnabled, false)
+
+	return img.ToString()
+}
+
+// OverrideAgentImage takes an existing image reference and potentially overrides portions of it based on the provided image configuration
+func OverrideAgentImage(currentImage string, overrideImageSpec *v2alpha1.AgentImageConfig) string {
+	image := FromString(currentImage)
+	overrideImage := fromImageConfig(overrideImageSpec)
```

Comment:
I think this overlooks the fact that [`AgentImageConfig.Name`](https://github.com/DataDog/datadog-operator/blob/main/api/datadoghq/v2alpha1/datadogagent_types.go#L1225-L1232) can be fully qualified name, or just name or name:tag. That's why tests are failing.